### PR TITLE
Add audio preprocessing stages (MonoConversion, SegmentConcatenation, TimestampMapper)

### PR DESCRIPTION
**Pull request by shubhamNvidia**
Adds foundational audio preprocessing stages for the ADV pipeline:
MonoConversionStage: Converts multi-channel audio to mono and verifies sample rate, with strict/non-strict sample rate enforcement.
SegmentConcatenationStage: Concatenates multiple audio segments with configurable silence gaps between them.
Config dataclasses (MonoConversionConfig, SegmentConcatenationConfig) with from_dict/to_dict support.
Unit tests with mocked I/O covering config, stage properties, process logic, and edge cases.

### CODE REVIEW

File: nemo_curator/stages/audio/configs/__init__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
+# Licensed under the Apache License, Version 2.0
+
+"""Configuration classes for audio processing stages."""
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/configs/__init__.py]: Let's use the full copyright here too.

File: nemo_curator/stages/audio/configs/concatenation.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# limitations under the License.
+
+"""Configuration for Segment Concatenation stage."""
+
+from dataclasses import dataclass, asdict
+from typing import Dict, Any
+
+
+@dataclass
+class SegmentConcatenationConfig:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/configs/concatenation.py]: Discussed offline. We do not need a config class per stage.

File: nemo_curator/stages/audio/configs/mono_conversion.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# limitations under the License.
+
+"""Configuration for Mono Conversion stage."""
+
+from dataclasses import dataclass, asdict
+from typing import Dict, Any
+
+
+@dataclass
+class MonoConversionConfig:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/configs/mono_conversion.py]: Discussed offline. We do not need a config class per stage.

File: nemo_curator/stages/audio/preprocessing/concatenation.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                'sample_rate': 48000,
+            }
+            
+            first_task = tasks[0]
+            return [AudioBatch(
+                data=output_data,
+                task_id="concatenated",
+                dataset_name=first_task.dataset_name,
+                _metadata=first_task._metadata,
+                _stage_perf=list(first_task._stage_perf),
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/preprocessing/concatenation.py]: Discussed offline. We should not use only the first task for this.

File: nemo_curator/stages/audio/preprocessing/concatenation.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            )]
+            
+        except Exception as e:
+            logger.error(f"Error concatenating segments: {e}")
+            return []
+    
+    def process(self, task: AudioBatch) -> Optional[AudioBatch]:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/preprocessing/concatenation.py]: Discussed offline. We should not need to implement `process` if we are implementing `process_batch`.

File: nemo_curator/stages/audio/preprocessing/concatenation.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        config = SegmentConcatenationConfig(silence_duration_sec=0.5)
+        stage = SegmentConcatenationStage(config=config)
+    """
+    
+    config: Optional[SegmentConcatenationConfig] = None
+    silence_duration_sec: float = 1.0
+    audio_key: str = "audio"
+    
+    name: str = "SegmentConcatenation"
+    batch_size: int = 10
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/preprocessing/concatenation.py]: Discussed offline. There is no way to ensure that the 10 batches included in a single `process_batch` are related to each other/correspond to each other in a specified way. This stage may need reworking.

File: tests/stages/audio/preprocessing/test_concatenation.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        cfg = SegmentConcatenationConfig.from_dict(
+            {"silence_duration_sec": 0.5, "batch_size": 20}
+        )
+        assert cfg.silence_duration_sec == 0.5
+        assert cfg.batch_size == 20
+        assert cfg.audio_key == "audio"
+
+    def test_from_dict_ignores_unknown_keys(self) -> None:
+        cfg = SegmentConcatenationConfig.from_dict({"foo": "bar"})
+        assert cfg.silence_duration_sec == 1.0
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_concatenation.py]: We do not need tests for checking default values, parameters being set, etc.

File: tests/stages/audio/preprocessing/test_concatenation.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        assert "num_segments" in output_keys
+        assert "total_duration_sec" in output_keys
+        assert "concatenated" in output_keys
+
+    def test_config_overrides_params(self) -> None:
+        cfg = SegmentConcatenationConfig(silence_duration_sec=2.5, audio_key="wav", batch_size=5)
+        stage = SegmentConcatenationStage(config=cfg)
+        assert stage.silence_duration_sec == 2.5
+        assert stage.audio_key == "wav"
+        assert stage.batch_size == 5
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_concatenation.py]: See above comment.

File: tests/stages/audio/preprocessing/test_concatenation.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+        stage = SegmentConcatenationStage(silence_duration_sec=2.0)
+        result = stage.process_batch(tasks)
+
+        combined = result[0].data[0]["audio"]
+        combined_duration_sec = len(combined) / 1000.0
+        expected = 1.0 + 2.0 + 1.0
+        assert abs(combined_duration_sec - expected) < 0.1
+
+    def test_process_single_delegates_to_batch(self) -> None:
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_concatenation.py]: This can be removed.

File: tests/stages/audio/preprocessing/test_concatenation.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+    def test_no_audio_segment_in_items(self) -> None:
+        tasks = [AudioBatch(data=[{"other_key": "value"}], dataset_name="ds")]
+
+        stage = SegmentConcatenationStage()
+        result = stage.process_batch(tasks)
+
+        assert result == []
+
+    def test_preserves_dataset_name(self) -> None:
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_concatenation.py]: This can be removed.

File: tests/stages/audio/preprocessing/test_mono_conversion.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        cfg = MonoConversionConfig.from_dict(
+            {"output_sample_rate": 16000, "strict_sample_rate": False}
+        )
+        assert cfg.output_sample_rate == 16000
+        assert cfg.strict_sample_rate is False
+        assert cfg.audio_filepath_key == "audio_filepath"
+
+    def test_from_dict_ignores_unknown_keys(self) -> None:
+        cfg = MonoConversionConfig.from_dict({"unknown_key": 42})
+        assert cfg.output_sample_rate == 48000
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_mono_conversion.py]: These can be removed.

File: tests/stages/audio/preprocessing/test_mono_conversion.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        stage = MonoConversionStage()
+        assert stage.name == "MonoConversion"
+        assert stage.inputs() == (["data"], [])
+        assert stage.outputs() == ([], ["waveform", "sample_rate", "is_mono", "duration", "num_samples"])
+
+    def test_config_overrides_params(self) -> None:
+        cfg = MonoConversionConfig(output_sample_rate=16000, strict_sample_rate=False)
+        stage = MonoConversionStage(config=cfg)
+        assert stage.output_sample_rate == 16000
+        assert stage.strict_sample_rate is False
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_mono_conversion.py]: These can be removed.

File: tests/stages/audio/preprocessing/test_mono_conversion.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+        with patch("nemo_curator.stages.audio.preprocessing.mono_conversion.sf.read", side_effect=RuntimeError("bad file")):
+            with patch("os.path.exists", return_value=True):
+                stage = MonoConversionStage()
+                batch = AudioBatch(data=[{"audio_filepath": wav.as_posix()}])
+                result = stage.process(batch)
+
+        assert len(result.data) == 0
+
+    def test_preserves_task_metadata(self, tmp_path: Path) -> None:
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_mono_conversion.py]: This can be removed.

File: nemo_curator/stages/audio/configs/__init__.py (hunk 1 above)
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/configs/__init__.py]: done

File: nemo_curator/stages/audio/__init__.py [hunk 14]
```diff
@@ -20,16 +20,37 @@
...
     "GetAudioDurationStage",
     "LegacySpeechStage",
+    "MonoConversionStage",
     "PreserveByValueStage",
-]
+    "SegmentConcatenationStage",
+    "TimestampMapperStage",
+    "MonoConversionConfig",
+    "SegmentConcatenationConfig",
+    "TimestampMapperConfig",
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/__init__.py]:
```suggestion
    "TimestampMapperConfig",
]
```

**Comment by sarahyurick** [comment]
Hi @shubhamNvidia can you please merge main and resolve the conflicts? Then I can kick off the Ruff linter which will show us all of the style updates needed.

**Comment by shubhamNvidia** [comment]
@sarahyurick I’ve merged with main and resolved the conflicts—you can go ahead and run the Ruff linter now

**Comment by sarahyurick** [comment]
Hi @shubhamNvidia the list of Ruff changes needed for this PR can be found here: https://github.com/NVIDIA-NeMo/Curator/actions/runs/23552031600/job/68568380509?pr=1575

**Comment by sarahyurick** [comment]
Hi @shubhamNvidia please update: https://github.com/NVIDIA-NeMo/Curator/actions/runs/23808747074/job/69389685007?pr=1575

File: tests/stages/audio/preprocessing/test_concatenation.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+            data={"segments": [{"other_key": "value"}]},
+            task_id="empty",
+            dataset_name="ds",
+        )
+        stage = SegmentConcatenationStage()
+        result = stage.process(task)
+        assert result == []
+
+    def test_missing_segments_key_raises(self) -> None:
+        import pytest
```
Comment (sarahyurick) [inline tests/stages/audio/preprocessing/test_concatenation.py]: Very nit: please move to top-level import.

Comment (shubhamNvidia) [reply tests/stages/audio/preprocessing/test_concatenation.py]: sure

Comment (shubhamNvidia) [reply tests/stages/audio/preprocessing/test_concatenation.py]: Done

**Review by sarahyurick — APPROVED** [review]
LGTM, thanks!
